### PR TITLE
Issue/11443 onboarding blaze error handling

### LINKED ISSUE #11443: [Dynamic Dashboard] Improve error handling for Blaze and Onboarding

### PR DESCRIPTION
**Pull request by hichamboushaba**
Closes: #11443 

### Description
This PR improves error handling in the Onboarding and Blaze cards.

### Testing instructions
1. Apply the following patch
<details>
<summary>Patch</summary>

```Patch
Subject: [PATCH] add
---
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/list/ProductListRepository.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/list/ProductListRepository.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/list/ProductListRepository.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/list/ProductListRepository.kt	(revision 28530588579adf71d6151cee9e4b29cab28e8237)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/list/ProductListRepository.kt	(date 1714822175705)
@@ -20,6 +20,7 @@
 import com.woocommerce.android.util.CoroutineDispatchers
 import com.woocommerce.android.util.WooLog
 import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.delay
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.map
 import kotlinx.coroutines.withContext
@@ -31,6 +32,7 @@
 import org.wordpress.android.fluxc.model.WCProductModel
 import org.wordpress.android.fluxc.store.WCProductStore
 import javax.inject.Inject
+import kotlin.random.Random
 
 class ProductListRepository @Inject constructor(
     private val appPrefsWrapper: AppPrefsWrapper,
@@ -87,6 +89,10 @@
         excludedProductIds: List<Long>? = null,
         sortType: WCProductStore.ProductSorting? = null
     ): Result<List<Product>> {
+        if (Random.nextBoolean()) {
+            delay(300)
+            return Result.failure(Exception())
+        }
         offset = if (loadMore) offset + PRODUCT_PAGE_SIZE else 0
         lastSearchQuery = null
         lastIsSkuSearch = WCProductStore.SkuSearchOptions.Disabled
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/onboarding/StoreOnboardingRepository.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/onboarding/StoreOnboardingRepository.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/onboarding/StoreOnboardingRepository.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/onboarding/StoreOnboardingRepository.kt	(revision 28530588579adf71d6151cee9e4b29cab28e8237)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/onboarding/StoreOnboardingRepository.kt	(date 1714822175707)
@@ -20,6 +20,7 @@
 import org.wordpress.android.fluxc.store.SiteStore.SiteVisibility.PUBLIC
 import javax.inject.Inject
 import javax.inject.Singleton
+import kotlin.random.Random
 
 @Singleton
 class StoreOnboardingRepository @Inject constructor(
@@ -39,6 +40,7 @@
     suspend fun fetchOnboardingTasks(): Result<Unit> {
         WooLog.d(WooLog.T.ONBOARDING, "Fetching onboarding tasks")
         val result = onboardingStore.fetchOnboardingTasks(selectedSite.get())
+        if (Random.nextBoolean()) return Result.failure(Exception())
         return when {
             result.isError -> {
                 WooLog.i(WooLog.T.ONBOARDING, "Error fetching onboarding tasks: ${result.error}")
Index: build.gradle
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/build.gradle b/build.gradle
--- a/build.gradle	(revision 28530588579adf71d6151cee9e4b29cab28e8237)
+++ b/build.gradle	(date 1714822175709)
@@ -30,7 +30,7 @@
     tasks.withType(KotlinCompile).all {
         kotlinOptions {
             jvmTarget = JavaVersion.VERSION_1_8
-            allWarningsAsErrors = true
+            allWarningsAsErrors = false
             freeCompilerArgs += [
                     "-opt-in=kotlin.RequiresOptIn",
                     "-Xjvm-default=all-compatibility",
```

</details>

2. Open the app.
3. Refresh the Blaze and Onboarding cards, they will fail randomly.
4. Confirm retrying will reload the cards

### Images/gif
[Screen_recording_20240505_170129.webm](https://github.com/woocommerce/woocommerce-android/assets/1657201/875a1382-37ca-47c5-a7ce-d1446158ef5a)

- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardFragment.kt [hunk 1]
```diff
@@ -69,7 +67,6 @@ class DashboardFragment :
     }
 
     private val dashboardViewModel: DashboardViewModel by viewModels()
-    private val storeOnboardingViewModel: StoreOnboardingViewModel by activityViewModels()
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardFragment.kt:72]: This part is not related to changes of the PR, I just noticed we missed removing this from the fragment.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/onboarding/DashboardOnboardingViewModel.kt [hunk 2]
```diff
@@ -40,79 +39,72 @@ import dagger.assisted.Assisted
...
+        onViewAllTapped = DashboardWidgetAction(
+            titleResource = R.string.store_onboarding_task_view_all_tasks,
+            action = ::viewAllClicked
         )
     )
-    val viewState = _viewState
 
-    private val refreshTrigger = MutableSharedFlow<RefreshEvent>(extraBufferCapacity = 1)
+    @OptIn(ExperimentalCoroutinesApi::class)
+    val viewState = merge(parentViewModel.refreshTrigger, refreshTrigger)
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/onboarding/DashboardOnboardingViewModel.kt:85]: I really like how you merged the trigger and view state flows here (no pun intended 😝)

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/list/ProductListRepository.kt [hunk 3]
```diff
@@ -87,23 +83,40 @@ class ProductListRepository @Inject constructor(
...
-            )
-            dispatcher.dispatch(WCProductActionBuilder.newFetchProductsAction(payload))
-        }
+    ): Result<List<Product>> {
+        offset = if (loadMore) offset + PRODUCT_PAGE_SIZE else 0
+        lastSearchQuery = null
+        lastIsSkuSearch = WCProductStore.SkuSearchOptions.Disabled
 
-        return getProductList(productFilterOptions, excludedProductIds)
+        return productStore.fetchProducts(
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/list/ProductListRepository.kt:91]: Nice!

**Review by 0nko — APPROVED** [review]
Nice job! 👍
